### ISSUE TITLE
Core packages

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#448]
Come up with a list of 'core' webbpl packages then add them as npm dependencies.

This came up in the lab meeting (May 20th) on documentation.

**Comment by null-a** [comment]
What's the motivation for this approach vs. just adding the useful stuff to webppl? One reason I can think of is cases like IO, were there's code that's useful on Node but not the browser (or vice versa). What else?

Would there be any value in keeping these things as packages but including them in the main repo? This adds more structure to the code than we have now, and the code can still be loaded/bundled conditionally, but documentation, testing and versioning might be easier to handle than if things are spread over several packages.

**Comment by hawkrobe** [comment]
Another reason I had in mind was to simplify the interface to functional utils like those in underscore. Right now we've reimplemented a bunch of functions like map and filter that don't work from underscore, but there are also a bunch of useful underscore functions we still want, like contains or

**Comment by stuhlmueller** [comment]
> Would there be any value in keeping these things as packages but including them in the main repo? This adds more structure to the code than we have now, and the code can still be loaded/bundled conditionally, but documentation, testing and versioning might be easier to handle than if things are spread over several packages.

Sounds reasonable if it avoids difficulties, but I don't quite see yet that it does:

- Versioning seems easier to me if packages are in separate repositories since we can more easily view only the history of changes relevant to that package.

- We're already building the distribution docs using a script. We could use the same script to copy over documentation from npm-installed packages. 

- For testing, we could try to make our test runner run the tests in each of the npm-installed packages. Ideally, this would happen in a way that is part of a coherent story for testing individual packages on their own.

Of these, testing seems most likely to cause difficulties, but it also seems worth it to me to explore whether we can make it work. If our core packages are tested and documented in the same way that we recommend for other packages, they will serve as instructive examples that we can point people to, and we will be more likely to notice weaknesses in our approach to packages.

**Comment by longouyang** [comment]
Proposal: we include the following packages as core using [npm's `peerDependency` mechanism](https://nodejs.org/en/blog/npm/peer-dependencies/):

- json
- fs
- csv
- dp
- editor
- viz

Also, I've found that the `webppl-` prefix for packages is annoying. How do folks feel about switching to `wp-`?

**Comment by null-a** [comment]
> there are also a bunch of useful underscore functions we want that aren't reimplemented

I agree -- the situation with underscore is confusing. The bit I'm not following is how fixing that requires 'core packages'. These seem pretty much orthogonal to me.

> Sounds reasonable if it avoids difficulties, but I don't quite see yet that it does:

My point is that splitting things up across repos/packages incurs overhead. The thing I'm not clear about is what we expect to get back in return. It seems pretty clear that the editor and viz tools (for example) are best handled as separate projects, but why does it makes sense to have e.g. dp as a 'core package' rather than just add that function to webppl?

(It's not that I don't expect there to be reasons, it's just I can't figure out what they are!)

**Comment by ngoodman** [comment]
i kind of agree with paul, that we should avoid incurring maintenance overhead involved with separate repos. i'd say things like editor and viz should probably be separate repos; things like fs and csv (that are only useful on command line) can be in main repo but should be packages or something so they are not always included; things like dp are ambiguous -- they should probably just be part of the main project... the reason i had suggested separating them is that dp is a bit dangerous (can't be applied to stochastic functions) and i wanted a way to make it clear that there is some 'caveat emptor' that doesn't apply to core webppl stuff....

overall, i don't think we need a uniform solution to this.
